extensions/aws/processors/PutS3Object.cpp

```diff
@@ -0,0 +1,495 @@
+  if (file_cred) {
+    logger_->log_info("AWS Credentials successfully set from file");
+    return file_cred.value();
+  }
+
+  auto service_cred = getAWSCredentialsFromControllerService(context);
+  if (service_cred) {
+    logger_->log_info("AWS Credentials successfully set from controller service");
+    return service_cred.value();
+  }
+
+  return minifi::utils::nullopt;
+}
+
+void PutS3Object::fillUserMetadata(const std::shared_ptr<core::ProcessContext> &context) {
+  const auto &dynamic_prop_keys = context->getDynamicPropertyKeys();
+  bool first_property = true;
+  for (const auto &prop_key : dynamic_prop_keys) {
+    std::string prop_value = "";
+    if (context->getDynamicProperty(prop_key, prop_value) && !prop_value.empty()) {
+      logger_->log_debug("PutS3Object: DynamicProperty: [%s] -> [%s]", prop_key, prop_value);
+      put_s3_request_params_.user_metadata_map.emplace(prop_key, prop_value);
+      if (first_property) {
+        user_metadata_ = prop_key + "=" + prop_value;
+        first_property = false;
+      } else {
+        user_metadata_ += "," + prop_key + "=" + prop_value;
+      }
+    }
+  }
+  logger_->log_debug("PutS3Object: User metadata [%s]", user_metadata_);
+}
+
+bool PutS3Object::setProxy(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::FlowFile> &flow_file) {
+  aws::s3::ProxyOptions proxy;
+  context->getProperty(ProxyHost, proxy.host, flow_file);
+  std::string port_str;
+  if (context->getProperty(ProxyPort, port_str, flow_file) && !port_str.empty() && !core::Property::StringToInt(port_str, proxy.port)) {
+    logger_->log_error("Proxy port invalid");
+    return false;
+  }
+  context->getProperty(ProxyUsername, proxy.username, flow_file);
+  context->getProperty(ProxyPassword, proxy.password, flow_file);
+  if (!proxy.host.empty()) {
+    s3_wrapper_->setProxy(proxy);
+    logger_->log_info("Proxy for PutS3Object was set.");
+  }
+  return true;
+}
+
+void PutS3Object::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) {
+  if (!context->getProperty(Bucket.getName(), put_s3_request_params_.bucket) || put_s3_request_params_.bucket.empty()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Bucket property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Bucket [%s]", put_s3_request_params_.bucket);
+
+  if (!context->getProperty(StorageClass.getName(), put_s3_request_params_.storage_class)
+      || put_s3_request_params_.storage_class.empty()
+      || STORAGE_CLASSES.find(put_s3_request_params_.storage_class) == STORAGE_CLASSES.end()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Storage Class property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Storage Class [%s]", put_s3_request_params_.storage_class);
+
+  std::string value;
+  if (!context->getProperty(Region.getName(), value) || value.empty() || REGIONS.find(value) == REGIONS.end()) {
```

Comment:
Just fancy: in case you use set, you can rely on count instead of comparing the result of find to end.

Reply:
Fixed in [e994558](https://github.com/apache/nifi-minifi-cpp/pull/915/commits/e994558ba8904488b0e3489d51236fb7d4e31ce1)